src/Autofac/ContainerBuilder.cs

```diff
@@ -234,28 +235,89 @@ private void Build(IComponentRegistry componentRegistry, bool excludeDefaultModu
-                RegisterDefaultAdapters(componentRegistry);
+            RegisterDefaultAdapters(componentRegistry, flags);
 
             foreach (var callback in _configurationCallbacks)
                 callback.Callback(componentRegistry);
         }
 
-        private void RegisterDefaultAdapters(IComponentRegistry componentRegistry)
+        private void RegisterDefaultAdapters(IComponentRegistry componentRegistry, DefaultServiceFlags flags)
         {
-            this.RegisterGeneric(typeof(KeyedServiceIndex<,>)).As(typeof(IIndex<,>)).InstancePerLifetimeScope();
-            componentRegistry.AddRegistrationSource(new CollectionRegistrationSource());
-            componentRegistry.AddRegistrationSource(new OwnedInstanceRegistrationSource());
-            componentRegistry.AddRegistrationSource(new MetaRegistrationSource());
-            componentRegistry.AddRegistrationSource(new LazyRegistrationSource());
-            componentRegistry.AddRegistrationSource(new LazyWithMetadataRegistrationSource());
-            componentRegistry.AddRegistrationSource(new StronglyTypedMetaRegistrationSource());
-            componentRegistry.AddRegistrationSource(new GeneratedFactoryRegistrationSource());
+            if ((flags & DefaultServiceFlags.Index) == DefaultServiceFlags.Index)
+                this.RegisterGeneric(typeof(KeyedServiceIndex<,>)).As(new[] { typeof(IIndex<,>) }).InstancePerLifetimeScope();
+            if ((flags & DefaultServiceFlags.Collections) == DefaultServiceFlags.Collections)
+                componentRegistry.AddRegistrationSource(new CollectionRegistrationSource());
+            if ((flags & DefaultServiceFlags.Owned) == DefaultServiceFlags.Owned)
+                componentRegistry.AddRegistrationSource(new OwnedInstanceRegistrationSource());
+            if ((flags & DefaultServiceFlags.Meta) == DefaultServiceFlags.Meta)
+                componentRegistry.AddRegistrationSource(new MetaRegistrationSource());
+            if ((flags & DefaultServiceFlags.Lazy) == DefaultServiceFlags.Lazy)
+                componentRegistry.AddRegistrationSource(new LazyRegistrationSource());
+            if ((flags & DefaultServiceFlags.LazyWithMeta) == DefaultServiceFlags.LazyWithMeta)
+                componentRegistry.AddRegistrationSource(new LazyWithMetadataRegistrationSource());
+            if ((flags & DefaultServiceFlags.StronglyTypedMeta) == DefaultServiceFlags.StronglyTypedMeta)
+                componentRegistry.AddRegistrationSource(new StronglyTypedMetaRegistrationSource());
+            if ((flags & DefaultServiceFlags.GeneratedFactory) == DefaultServiceFlags.GeneratedFactory)
+                componentRegistry.AddRegistrationSource(new GeneratedFactoryRegistrationSource());
         }
 
         private List<Action<IContainer>> GetBuildCallbacks()
         {
             return (List<Action<IContainer>>)Properties[BuildCallbackPropertyKey];
         }
     }
+
+    [Flags]
+    public enum DefaultServiceFlags
```

Comment:
No problem with the concept but could do with a better name. Perhaps something like `FeatureToggles`.

Reply:
Or `ContainerFeatures`.

If this is intended to be a more granular way to control what gets built into the container, I'd recommend also including `StartableComponents` as a flag so it can do everything `ContainerBuildOptions` can do and more.

Reply:
It should also be in its own file, not attached to the bottom of `ContainerBuilder.cs`.